File: TorrentAdder/SYResultCell.m
//
//  SYResultCell.m
//  TorrentAdder
//
//  Created by syan on 9/28/15.
//  Copyright © 2015 Syan. All rights reserved.
//

#import "SYResultCell.h"
#import "SYWebAPI.h"
#import "NSDate+TimeAgo.h"
#import "NSAttributedString+SYKit.h"

@interface SYResultCell ()
@property (nonatomic, weak) IBOutlet UILabel *labelName;
@end

@implementation SYResultCell

+ (NSAttributedString *)attributedTitleForResult:(SYResultModel *)result
{
    NSMutableAttributedString *title = [[NSMutableAttributedString alloc] init];
    [title sy_appendString:[result.name stringByAppendingString:@"\n"]
                      font:[UIFont systemFontOfSize:16]
                     color:[UIColor darkTextColor]];
    
    [title sy_appendString:[result.size stringByAppendingString:@", "]
                      font:[UIFont systemFontOfSize:14]
                     color:[UIColor grayColor]];
    
    NSString *dateString = result.age;
    if (result.parsedDate)
    {
        NSTimeInterval aboutTwoMonths = 1440.*3600.;
        dateString = [result.parsedDate timeAgoWithLimit:aboutTwoMonths
                                              dateFormat:NSDateFormatterMediumStyle
                                           andTimeFormat:NSDateFormatterNoStyle];
    }
    
    UIFont *dateFont = [UIFont systemFontOfSize:14];
    if (result.parsedDate && fabs(result.parsedDate.timeIntervalSinceNow) < 48.*3600.)
    {
        dateFont = [UIFont systemFontOfSize:14 weight:UIFontWeightBold];
    }
    else if (result.parsedDate && fabs(result.parsedDate.timeIntervalSinceNow) < 360.*3600.)
    {
        dateFont = [UIFont systemFontOfSize:14 weight:UIFontWeightSemibold];
    }

    [title sy_appendString:[dateString stringByAppendingString:@", " ]
                      font:dateFont
                     color:[UIColor grayColor]];
    
    [title sy_appendString:[NSString stringWithFormat:@"%ld", (long)result.seed]
                      font:[UIFont systemFontOfSize:14]
                     color:[UIColor colorWithRed:0 green:0.56 blue:0.05 alpha:1.]];
    
    [title sy_appendString:@"/"
                      font:[UIFont systemFontOfSize:14]
                     color:[UIColor grayColor]];
    
    [title sy_appendString:[NSString stringWithFormat:@"%ld", (long)result.leech]
                      font:[UIFont systemFontOfSize:14]
                     color:[UIColor redColor]];
    
    return [title copy];
}

- (void)setResult:(SYResultModel *)result
{
    self->_result = result;
    [self.labelName setAttributedText:[[self class] attributedTitleForResult:result]];
}

+ (CGFloat)cellHeightForResult:(SYResultModel *)result
                         width:(CGFloat)width
{
    UILabel *label = [[UILabel alloc] init];
    [label setAttributedText:[self attributedTitleForResult:result]];
    [label setNumberOfLines:0];
    return [label sizeThatFits:CGSizeMake(width - 2 * 15, 1000)].height + 20;
}

@end

File: TorrentAdder/SYResultCell.swift
//
//  SYResultCell.swift
//  TorrentAdder
//
//  Created by Stanislas Chevallier on 28/11/2018.
//  Copyright © 2018 Syan. All rights reserved.
//

import UIKit
import NSDate_TimeAgo
import SYKit

class SYResultCell: UITableViewCell {

    // MARK: Properties
    var result: SYSearchResult? {
        didSet {
            updateContent()
        }
    }
    
    // MARK: Views
    @IBOutlet var label: UILabel!
    
    // MARK: Content
    private func updateContent() {
        guard let result = result else {
            label.text = nil
            return
        }
        
        var dateString = result.age ?? ""
        var dateFont = UIFont.systemFont(ofSize: 14)
        if let date = result.parsedDate {
            let twoMonths: TimeInterval = 1440 * 3600
            dateString = (date as NSDate).timeAgo(withLimit: twoMonths, dateFormat: .medium, andTimeFormat: .none)
            
            if fabs(date.timeIntervalSinceNow) < 48 * 3600 {
                dateFont = UIFont.systemFont(ofSize: 14, weight: .bold)
            }
            else if fabs(date.timeIntervalSinceNow) < 360 * 3600 {
                dateFont = UIFont.systemFont(ofSize: 14, weight: .semibold)
            }
        }
        
        let string = NSMutableAttributedString()
        
        string.append(result.name + "\n", font: UIFont.systemFont(ofSize: 16), color: .text)
        if let size = result.size {
            string.append(size + ", ", font: UIFont.systemFont(ofSize: 15), color: .subtext)
        }
        string.append(dateString + ", ", font: dateFont, color: .subtext)
        string.append(String(result.seed), font: UIFont.systemFont(ofSize: 14), color: .seeder)
        string.append("/", font: UIFont.systemFont(ofSize: 14), color: .gray)
        string.append(String(result.leech), font: UIFont.systemFont(ofSize: 14), color: .leechers)
        if result.verified {
            string.append(" ✔️", font: UIFont.systemFont(ofSize: 12), color: nil)
        }
        
        label.attributedText = string
    }
}
